### DISCUSSION TITLE
Checking whether a given content type or `Lang` can be used to read/write a certain Jena model class

### DISCUSSION THREAD

**Discussion by namedgraph** [apache/jena#2414] [General]
I'm trying to find a `ResultSet`-based equivalent to the `RDFLanguages` class but not able to.

There is `RDFParserRegistry` and `RDFWriterRegistry`. There is also `ResultSetReaderRegistry` and `ResultSetWriterRegistry`.
But there is only [`RDFLanguages`](https://jena.apache.org/documentation/javadoc/arq/org.apache.jena.arq/org/apache/jena/riot/RDFLanguages.html), not `ResultSetLanguages`. That seems inconsistent and asymmetric. Would it make sense to add such a class?

More specifically, what can I use as an equivalent of `RDFLanguages.contentTypeToLang()` but for `ResultSet` langs?

**Comment by rvesse** [comment]
See [`ResultSetLang`](https://github.com/apache/jena/blob/main/jena-arq/src/main/java/org/apache/jena/riot/resultset/ResultSetLang.java)

**Reply by namedgraph** [reply]
But that is like [`Lang`](https://github.com/apache/jena/blob/main/jena-arq/src/main/java/org/apache/jena/riot/Lang.java) :) A media type definition together with their instances, but not their registry.

**Comment by afs** [comment]
Use `RDFLanguages.contentTypeToLang`.

`ResultSetLang` has constants which are `Lang` and these are registered in `RDFLanguages`.

`ResultSetReaderRegistry.register` takes a `Lang`. There are no `ResultSetLang` objects.

```java
    Lang x = RDFLanguages.contentTypeToLang("application/sparql-results+xml");
    System.out.println(x);
```

**Reply by namedgraph** [reply]
Thanks @afs. But I have a follow-up question :) What is the canonical way to check that a `Lang` is one of `ResultSetLang`s (without hardcoding them)?

**Reply by afs** [reply]
`Lang` are _symbols_ (unique token; typed and not strings) together with HTTP information.

Ask the question as "_Can I use this `Lang` for_ <_purpose_>?".

"text" for example, is only for writing.

**Reply by namedgraph** [reply]
I feel like I'm struggling to explain myself 😄 

I need a method that checks whether a given content type or `Lang` can be used to read/write a certain Jena model class.

- for `Model`, there's `RDFLanguages::isTriples`
- for `Dataset`, there's `RDFLanguages::isQuads`
- for `ResultSet`, there is none?

Do you see what I mean? Unless there's a better way to do it altogether.

**Reply by namedgraph** [reply]
OK wait, I think I figured it out:
* for `Model`: `RDFParserRegistry::isTriples` and something from `RDFWriterRegistry`?
* for `Dataset`: `RDFParserRegistry::isQuads` and something from `RDFWriterRegistry`?
* for `ResultSet`: `ResultSetReaderRegistry::isRegistered` and `ResultSetWriterRegistry::isRegistered`

That looks like a better approach overall, but then `RDFWriterRegistry` has no methods to distiguish between triple and quad `Lang`s?

**Reply by afs** [reply]
triples/quads isn't an aspect of a write format (though it might be for some custom `RDFFormat`). It applies to both reading and writing. Ask `RDFLanguages` (arguable, remove from `RDFParserRegistry`).

You can write triples to quad formats.
JSON-LD confuses the situation - it is commonly talked about as  graph syntax but it is really a quad syntax.

`ResultSets` are not split by triple/quad capabilities. Either it is registered (has a factory) for a use or not.

**Reply by namedgraph** [reply]
> triples/quads isn't an aspect of a write format (though it might be for some custom `RDFFormat`). It applies to both reading and writing. Ask `RDFLanguages` (arguable, remove from `RDFParserRegistry`).

Andy I understand what you're saying, but it's not really what I'm asking about. Let me explain my use cases:
* building an `Accept` header value. Here I need a registry of readable triple/quad/result set langs because I want to prioritise langs in each group differently.
* in JAX-RS response body writing happens in providers that depend on the return object type, e.g. `MessageBodyReader<Model>`, `MessageBodyReader<Dataset>` or `MessageBodyReader<ResultSet>`. Each of those providers needs to know which langs can be used based on the generics class. Again it requires a registry of readable/writable types that apply to each class.

> You can write triples to quad formats.

I understand this, but not sure how it's related to my use case. If quads were requested, JAX-RS will write a `Dataset` and needs to use a quad format. Returning a triple format (even if there are no named graphs in the dataset) would violate the conneg.

> `ResultSets` are not split by triple/quad capabilities. Either it is registered (has a factory) for a use or not.

Yes this is clear as well, I don't think I was trying to claim `ResultSet` is related to quads or triples.

**Comment by namedgraph** [comment]
To summarise, my main 2 quibbles are:
* the classes `RDFParserRegistry`, `RDFWriterRegistry`, `ResultSetReaderRegistry`, `ResultSetWriterRegistry` are all `Lang` registries but are not aligned. Would it be reasonable for them to implement a single interface such as
  ```java
  interface LangRegistry {
    Collection<Lang> registeredLangs();
  }
  ```
  ? If you agree, I can create an issue.
* `RDFLanguages::isTriples` and `RDFLanguages::isQuads`, but no `RDFLanguages::isResultSet`. And why does `RDFLanguages` need a registry if parsers and writers already have their own `Lang` registries?

**Reply by afs** [reply]
> `interface LangRegistry`

They are static classes - not interfaces. We have just released Jena5; changing to objects is a breaking change.

If you want parallel capabilities via objects with `LangRegistry` interface, build it. Add the setup code as `ServiceLoader` and initialize after Jena. That could also add the "Accept" prioritization.

`RDFLanguages::isResultSet` could be done - make it a call to (new) `ResultSetLang.isRegistered` (+machinery).

> I can create an issue.

Are you going to provide a PR?

**Reply by namedgraph** [reply]
I can try :) Maybe I could add singleton objects but keep the static methods.

**Comment by afs** [comment]
Seems over-engineering.

* Add `RDFLanguages::isResultSet` - make it a call to (new) `ResultSetLang.isRegistered`.
* `interface LangRegistry` is a functional interface so implement three functions (one liners) calling the appropriate `RDFLanguages::is?????` function.
* Pass the right function to code that needs a general lookup algorithm specialised to a chosen context.

**Comment by namedgraph** [comment]
@afs seems like this has been addressed by #2510? Thank you for that :)

I still feel like a `contentTypeToLang` method for `ResultSetLang` is missing though?

**Comment by namedgraph** [comment]
Ah OK, I see they're part of `RDFLanguages` now - which was not the case in 4.7.0:
https://github.com/apache/jena/pull/2510/files#diff-4efee2e599cd106d504a1687b6fff298e6fb0b24d41e4db707d88088f76d2afcR102
